### PR TITLE
coef names

### PR DESCRIPTION
**Pull request by aviatesk**
addresses https://github.com/alan-turing-institute/MLJLinearModels.jl/issues/60

requires:

- [x]  https://github.com/alan-turing-institute/MLJModelInterface.jl/pull/39 to be merged

/cc @tlienart @ablaom

### CODE REVIEW

File: src/mlj/interface.jl [hunk 1]
```diff
@@ -73,20 +74,24 @@ function MMI.predict(m::Union{CLF_MODELS...}, (θ, c, classes), Xnew)
     return [MMI.UnivariateFinite(classes, preds[i, :]) for i in 1:size(Xmatrix,1)]
 end
 
-function MMI.fitted_params(m::Union{CLF_MODELS...}, (θ, c, classes))
+function MMI.fitted_params(m::Union{CLF_MODELS...}, (θ, features, c, classes))
     if c > 1
+        W = reshape(θ, :, c)
         if m.fit_intercept
-            W = reshape(θ, div(length(θ), c), c)
-            return (coefs = W, intercept = nothing)
```
Comment (aviatesk) [inline src/mlj/interface.jl:80]: I _think_ the previous implementation was wrong -- if `m.fit_intercept` is true we should return `intercept`, no ?

File: src/mlj/interface.jl [hunk 2]
```diff
@@ -73,20 +74,24 @@ function MMI.predict(m::Union{CLF_MODELS...}, (θ, c, classes), Xnew)
...
+            return (coefs = coef_dict(W, features), intercept = W[end, :])
         end
-        W = reshape(θ, p+1, c)
-        return (coefs = W[1:p, :], intercept = W[end, :])
+        return (coefs = coef_dict(W[1:end-1,:], features), intercept = nothing)
```
Comment (aviatesk) [inline src/mlj/interface.jl:83]: The previous implementation actually errors in this pass (`p` is undefined). Let me know I introduce another error :P

File: src/mlj/interface.jl [hunk 3]
```diff
@@ -73,20 +74,24 @@ function MMI.predict(m::Union{CLF_MODELS...}, (θ, c, classes), Xnew)
...
-    m.fit_intercept && return (coefs = θ[1:end-1], intercept = θ[end])
-    return (coefs = θ, intercept = nothing)
+    m.fit_intercept && return (coefs = coef_dict(θ[1:end-1], features), intercept = θ[end])
+    return (coefs = coef_dict(θ, features), intercept = nothing)
 end
 
+coef_dict(W::AbstractMatrix, features) = Dict(feature => coef for (feature, coef) in zip(features, eachrow(W)))
+coef_dict(θ::AbstractVector, features) = Dict(feature => coef for (feature, coef) in zip(features, θ))
+coef_dict(W::AbstractMatrix, ::Nothing) = W
+coef_dict(W::AbstractVector, ::Nothing) = W
```
Comment (aviatesk) [inline src/mlj/interface.jl:93]: verbosity to avoid method ambiguity error ...

**Comment by aviatesk** [comment]
So in order for CIs to pass without any failure, we have to wait those PRs to be merged and new versions to be tagged for each package:
- https://github.com/alan-turing-institute/MLJBase.jl/pull/280
- https://github.com/alan-turing-institute/MLJModelInterface.jl/pull/39

Then we can just bump up the compat section for MLJModelInterface.jl

**Comment by aviatesk** [comment]
I will bump the MLJModelInterface.jl compat once the new version of MLJBase is released.
Then the compats would be all good and we can check how CI goes. (maybe after that I will add a test on this)

/cc @ablaom

**Comment by ablaom** [comment]
@aviatesk I think can update the [compat] now and retrigger CI

**Comment by aviatesk** [comment]
nice, thanks for notifying the new release.

**Comment by aviatesk** [comment]
okay, so now `fittted_params` returns:

> binary class classification
```julia
julia> fitted_params(mach)
(classes = CategoricalString{UInt32}["B", "O"],
 coefs = [:FL => 2.6201486950992203, :RW => 0.39292726376244974, :CL => 0.42944513973557696, :CW => -2.364090952736262, :BD => 2.184956788698077],
 intercept = -3.805377536908276,)
```

> multi class classification:
```julia
julia> fitted_params(mach)
(classes = CategoricalString{UInt32}["setosa", "versicolor", "virginica"],
 coefs = Pair{Symbol,SubArray{Float64,1,Array{Float64,2},Tuple{Int64,Base.Slice{Base.OneTo{Int64}}},true}}[:sepal_length => [0.2059480869774525, 0.15620597909667314, -0.3621540660741237], :sepal_width => [1.329471903695349, -0.5634731456659896, -0.765998758029362], :petal_length => [-2.4826936165327167, -0.1984673279849103, 2.6811609445176328], :petal_width => [-1.1274927833807755, -0.9013966540870679, 2.0288894374678437]],
 intercept = [0.22683530315174644, -0.2268353017119584, -15.374523249617724],)
```

> regression:
```julia
julia> fitted_params(mach)
(coefs = [:Crim => -0.09703973833125352, :Zn => 0.09842256842389137, :Indus => 0.05193156265302136, :NOx => 0.08164973034302989, :Rm => 1.9136770861182466, :Age => 0.08012259423475827, :Dis => -0.1673060345082267, :Rad => 0.2033714467207396, :Tax => -0.012765891530541989, :PTRatio => 0.5304063436035396, :Black => 0.0199836811158726, :LStat => -0.7721969054676527],
 intercept = 0.17859876246704648,)
```

where `coefs` now returns vector of pairs of feature name and coef when given train data contains `schema` information, otherwise returns just coefs as before.

It now also returns `classes` field for classification models, which enables users to see which coef corresponds to which class.

***

Possible discussion would be that we may want to keep `coefs` fields as just vector of coefs and keep those feature names as a separate field (say, `features`), as `classes` are so.
That way, `fitted_params(some_model)` would be always `Vector{Float64}` whenever given train data contains `schema` information, and would be more consistent, but then users have to match feature name to each coef manually, e.g. for plotting.
I'm not sure which way we want to go, and so very open to hear your ideas.

**Comment by aviatesk** [comment]
> Possible discussion would be that we may want to keep coefs fields as just vector of coefs and keep those feature names as a separate field (say, features), as classes are so.

what do you think on this, @tlienart ?

**Comment by tlienart** [comment]
sorry this is fine but fails on nightly due to some weird unrelated issue; could I ask you to allow failure on nightly in the travis yml file? thanks!

**Comment by aviatesk** [comment]
okay then let's go with the current implementation. 

> could I ask you to allow failure on nightly in the travis yml file?
done 👍

**Comment by tlienart** [comment]
Thanks

**Pull request merged**